### PR TITLE
Expr: Array

### PR DESCRIPTION
**Pull request by HoshinoTented**
- [x] Producer for ARRAY_EXPR
- [x] ExprResolver for Expr.Array
- [x] Desugarer for Expr.Array
- [x] Pretty Print for Expr.Array (I am not sure)

#552

### CODE REVIEW

File: base/src/main/java/org/aya/distill/ConcreteDistiller.java [hunk 1]
```diff
@@ -154,6 +154,20 @@ yield visitCalls(false, fn, (nc, l) -> l.toDoc(options), outer,
           .map(app -> term(Outer.Free, app)))
       );
       case Expr.Do aDo -> throw new UnsupportedOperationException("TODO");
```
Comment (ice1000) [inline base/src/main/java/org/aya/distill/ConcreteDistiller.java:156]: Can you implement this too?

File: base/src/main/java/org/aya/resolve/visitor/ExprResolver.java [hunk 2]
```diff
@@ -116,6 +116,47 @@ public record ExprResolver(
         if (type == path.type() && par == path.partial()) yield path;
         yield new Expr.Path(path.sourcePos(), path.params(), type, par);
       }
+      case Expr.Array arrayExpr -> arrayExpr.arrayBlock().fold(
+        left -> {
+          var bindName = resolve(left.bindName(), ctx);
+          var pureName = resolve(left.pureName(), ctx);
+
+          // region The same coding as Expr.Do! 😭
```
Comment (ice1000) [inline base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:124]: Maybe you can try generalizing the code?

File: base/src/test/java/org/aya/concrete/DesugarTest.java [hunk 3]
```diff
@@ -54,6 +54,16 @@ public class DesugarTest {
       }""");
   }
 
+  @Test public void arrays() {
+    desugarAndPretty("""
+    def simpleList => [ 1, 2, 3 ]
+    def listGenerator => [ x * y | x <- xs, y <- ys ]
```
Comment (ice1000) [inline base/src/test/java/org/aya/concrete/DesugarTest.java:60]: I guess you need to format this code

File: base/src/test/java/org/aya/concrete/DesugarTest.java [hunk 4]
```diff
@@ -54,6 +54,16 @@ public class DesugarTest {
...
+    """, """
+    def simpleList => :< 1 (:< 2 (:< 3 nil))
+    def listGenerator => >>= xs (\\ (x : _) => >>= ys (\\ (y : _) => pure (x * y)))
+    """);   // Aya doesn't know the `:<` and `>>=` are infix function in this case, how to let she know?
```
Comment (ice1000) [inline base/src/test/java/org/aya/concrete/DesugarTest.java:64]: I think you have to define those functions. Write something like `def infix >>= => {??}` I guess

**Review by ice1000 — COMMENTED** [review]
Rest looks good

**Comment by ice1000** [comment]
I also think `desugarAndPretty` is wrongly implemented. It doesn't resolve. Can you fix it in another PR? Like adding resolve and fix those tests who fails by adding unresolved definitions

File: base/src/test/java/org/aya/concrete/DesugarTest.java (hunk 3 above)
Comment (ice1000) [reply base/src/test/java/org/aya/concrete/DesugarTest.java:60]: No, I think it should be indented

File: base/src/test/java/org/aya/concrete/DesugarTest.java [hunk 5]
```diff
@@ -54,6 +54,27 @@ public class DesugarTest {
...
+      data List (A : Type 0)
+        | nil
+        | :< (_ : A) (_ : List A)
+      open List hiding ()
+      def >>= => {??}
+      def simpleList => :< 1 (:< 2 (:< 3 nil))
+      def listGenerator => >>= xs (\\ (x : _) => >>= ys (\\ (y : _) => pure (x * y)))
+      """); // TODO: make desugarAndPretty does resolve [#556](https://github.com/aya-prover/aya-dev/pull/556#issuecomment-1268376917).
+  }
+
```
Comment (imkiva) [inline base/src/test/java/org/aya/concrete/DesugarTest.java:77]: I would suggest testing the desugar in some `.aya` files, like the code below that tests the BinOp desugar:

https://github.com/aya-prover/aya-dev/blob/5e317b38ea27dbe878a1f13369102ca94e6044eb/base/src/test/resources/success/src/Assoc.aya#L13-L25

File: cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java [hunk 6]
```diff
@@ -604,13 +604,55 @@ public TeleDecl.FnDecl fnDecl(@NotNull GenericNode<?> node, Stmt.Accessibility a
           })
           .toImmutableSeq());
     }
-    // TODO: implement this
     if (node.is(ARRAY_EXPR)) {
-      return new Expr.HoleExpr(pos, false, null);
+      var arrayBlock = node.peekChild(ARRAY_BLOCK);
+
+      if (arrayBlock != null) {
+        if (arrayBlock.is(ARRAY_COMP_BLOCK)) {
```
Comment (imkiva) [inline cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java:611]: - extract two helper methods: `arrayCompBlock(GenericNode)` and `arrayElementList(GenericNode)`
- return early when `arrayBlock == null`

File: cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java [hunk 7]
```diff
@@ -604,13 +604,55 @@ public TeleDecl.FnDecl fnDecl(@NotNull GenericNode<?> node, Stmt.Accessibility a
...
+
     return unreachable(node);
   }
 
+  /**
+   * This function assumed that the node is DO_BINDING
+   */
+  public @NotNull Expr.DoBind buildDoBinding(@NotNull GenericNode<?> node) {
+    var wp = weakId(node.child(WEAK_ID));
+    return new Expr.DoBind(wp.sourcePos(), new LocalVar(wp.data()), expr(node.child(EXPR)));
```
Comment (imkiva) [inline cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java:653]: Use `LocalVar.from(wp)`

File: base/src/main/java/org/aya/resolve/visitor/ExprResolver.java (hunk 2 above)
Comment (ice1000) [reply base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:124]: No it's not. Your version is better.

File: cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java (hunk 6 above)
Comment (imkiva) [reply cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java:610]: Yes and braces can be omitted like
```
if (arrayBlock == null) return ...;
if (arrayBlock.is(...)) return arrayCompBlock(arrayBlock);
if (arrayBlock.is(...)) return arrayElementList(arrayBlock);

return unreachable(arrayBlock);
```

**Review by ice1000 — APPROVED** [review]
Cool!!
bors r+

**Pull request merged**